packages/stargate/src/queries/queryclient.ts

```diff
@@ -239,28 +239,26 @@ export class QueryClient {
     }
 
     if (headersSubscription) {
-      // get the header for height+1
-      nextHeader = await firstEvent(headersSubscription); // TODO: fall back on polling if this returns a too high header
-    } else {
+      const firstHeader = await firstEvent(headersSubscription);
+      // The first header we get might not be n+1 but n+2 or even higher. In such cases we fall back on a query.
+      if (firstHeader.height === searchHeight) {
+        nextHeader = firstHeader;
+      }
+    }
+
+    while (!nextHeader) {
       // start from current height to avoid backend error for minHeight in the future
-      let header = (await this.tmClient.blockchain(height, searchHeight)).blockMetas
+      const correctHeader = (await this.tmClient.blockchain(height, searchHeight)).blockMetas
```

Comment:
You could just directly assign this to `nextHeader` since in the fail case `find()` returns `undefined`.

Reply:
hmm, yeah, that could be done. However, I think it is easier to assign `nextHeader` only once when found.